server/db/driver/pg/accounts.go

```diff
@@ -158,37 +144,62 @@ func (a *Archiver) PayAccount(aid account.AccountID, coinID []byte) error {
 	return nil
 }
 
-// Get the next address for the current master pubkey.
-func (a *Archiver) getNextAddress() (string, error) {
-	stmt := fmt.Sprintf(internal.IncrementKey, feeKeysTableName)
-	var childExtKey *hdkeychain.ExtendedKey
-out:
-	for {
-		var child uint32
-		err := a.db.QueryRow(stmt, a.keyHash).Scan(&child)
-		if err != nil {
-			return "", err
-		}
-		childExtKey, err = a.feeKeyBranch.Child(child)
-		switch {
-		case errors.Is(err, hdkeychain.ErrInvalidChild):
-			continue
-		case err == nil:
-			break out
-		default:
-			log.Errorf("error creating child key: %v", err)
-			return "", fmt.Errorf("error generating fee address")
-		}
+// createKeyEntry creates an entry for the pubkey (hash) if it doesn't already
+// exist, and returns the child index.
+func createKeyEntry(db *sql.DB, tableName string, keyHash []byte) (uint32, error) {
+	stmt := fmt.Sprintf(internal.InsertKeyIfMissing, tableName)
+	var child uint32
+	err := db.QueryRow(stmt, keyHash).Scan(&child)
+	if err != nil {
+		return 0, err
 	}
+	return child, nil
+}
 
-	pubKeyBytes := childExtKey.SerializedPubKey()
-	addr, err := stdaddr.NewAddressPubKeyHashEcdsaSecp256k1V0(stdaddr.Hash160(pubKeyBytes), a.keyParams)
-	if err != nil {
-		log.Errorf("Failed to create creating new pubkey hash address: %v", err)
-		return "", fmt.Errorf("error encoding fee address: %w", err)
+// CreateFeeKeyEntryFromPubKey an entry for the pubkey, which should be of
+// length secp256k1.PubKeyBytesLenCompressed. Decred's HASH160 will be applied
+// internally. The current child index is returned.
+func (a *Archiver) CreateFeeKeyEntryFromPubKey(xpub string) (child uint32, err error) {
+	keyHash := dcrutil.Hash160([]byte(xpub))
+	log.Debugf("Inserting key entry for xpub %s, hash160 = %x", xpub, keyHash)
+	return createKeyEntry(a.db, a.tables.feeKeys, keyHash)
+}
+
+// FeeKeyIndex returns the current child index for the an xpub. If it is not
+// known, this returns an ErrUnknownFeeKey coded ArchiveError.
+func (a *Archiver) FeeKeyIndex(xpub string) (uint32, error) {
+	keyHash := dcrutil.Hash160([]byte(xpub))
+	stmt := fmt.Sprintf(internal.CurrentKeyIndex, feeKeysTableName)
+	var child uint32
+	err := a.db.QueryRow(stmt, keyHash).Scan(&child)
+	switch {
+	case errors.Is(err, sql.ErrNoRows):
+		return 0, db.ArchiveError{Code: db.ErrUnknownFeeKey}
+	case err == nil:
+	default:
+		return 0, err
 	}
+	return child, nil
+}
 
-	return addr.String(), nil
+// SetFeeKeyIndex records the child index for an xpub. An error is returned
+// unless exactly 1 row is updated.
+func (a *Archiver) SetFeeKeyIndex(idx uint32, xpub string) error {
+	keyHash := dcrutil.Hash160([]byte(xpub))
+	log.Debugf("Recording new index %d for xpub %s (%x)", idx, xpub, keyHash)
```

Comment:
How do you feel about truncating and just showing the first 20ish letters of the pubkey base58. For one it is very long and I doubt all needed to compare two pubkeys
```
[DBG] DB: Recording new index 1 for xpub tpubD6NzVbkrYhZ4XgiXtGrdW5XDAPFCL9h7we1vwNCpn8tGbBcgfVYjXyhWo4E1xkh56hjod1RhGjxbaTLV3X4FyWuejifB9jusQ46QzG87VKp (450907c35b4fa2c399bde54b1511aa48f5e4c3da)
```
And another, maybe it's not ideal to have it all over the logs. I would say it is sensitive info. Not super sensitive, but could be used in some sort of 300IQ attack.

Reply:
Got a point there.  I had figured if you have the control over the db you have control over the logger, but risk leaking data on yet another channel.  It's only public key, but still...